### PR TITLE
hwmap: add Mediacodec/Vulkan support

### PR DESCRIPTION
**Pull request by mbouron-gpsw**

### CODE REVIEW

File: libnodegl/backends/vk/pipeline_vk.c [hunk 1]
```diff
@@ -695,6 +772,28 @@ int ngli_pipeline_vk_update_texture(struct pipeline *s, int index, const struct
...
 
+    if (texture) {
+        struct texture_vk *texture_vk = (struct texture_vk *)texture;
+        if (need_desc_set_layout_update(s, texture_binding, texture)) {
+            VkDescriptorSetLayoutBinding *desc_bindings = ngli_darray_data(&s_priv->desc_set_layout_bindings);
+            VkDescriptorSetLayoutBinding *desc_binding = &desc_bindings[texture_binding->desc_binding_index];
+            texture_binding->use_ycbcr_sampler = texture_vk->use_ycbcr_sampler;
+            ngli_ycbcr_sampler_vk_unrefp(&texture_binding->ycbcr_sampler);
+            if (texture_vk->use_ycbcr_sampler) {
+                texture_binding->ycbcr_sampler = ngli_ycbcr_sampler_vk_ref(texture_vk->ycbcr_sampler);
```
Comment (cboesch-gpsw) [inline libnodegl/backends/vk/pipeline_vk.c:783]: I think you need to explain somewhere the purpose of the ref counting and the related lifetime considerations.

File: libnodegl/backends/vk/ycbcr_sampler_vk.c [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        .components    = params->components,
+        .xChromaOffset = params->x_chroma_offset,
+        .yChromaOffset = params->y_chroma_offset,
+        .chromaFilter  = params->filter,
+        .forceExplicitReconstruction = VK_FALSE,
+    };
+
+    VkResult res = vk->CreateSamplerYcbcrConversionKHR(vk->device, &sampler_ycbcr_info, 0, &s->conv);
+    if (res != VK_SUCCESS) {
+        LOG(ERROR, "could not create sampler ycbcr conversion: %s", ngli_vk_res2str(res));
```
Comment (cboesch-gpsw) [inline libnodegl/backends/vk/ycbcr_sampler_vk.c:68]: nit: YCbCr

File: libnodegl/meson.build [hunk 3]
```diff
@@ -398,6 +399,7 @@ gbackends_cfg = {
         'frameworks': ['Metal'],
       },
       'android': {
+        'src': files('backends/vk/hwmap_mediacodec_vk.c'),
```
Comment (cboesch-gpsw) [inline libnodegl/meson.build:402]: Add a changelog entry maybe?

**Review by cboesch-gpsw — CHANGES_REQUESTED** [review]

File: libnodegl/backends/vk/vkcontext.c [hunk 4]
```diff
@@ -622,6 +622,7 @@ static VkResult create_device(struct vkcontext *s)
         VK_EXT_EXTERNAL_MEMORY_DMA_BUF_EXTENSION_NAME,
         VK_EXT_IMAGE_DRM_FORMAT_MODIFIER_EXTENSION_NAME,
         VK_GOOGLE_DISPLAY_TIMING_EXTENSION_NAME,
+        VK_KHR_SAMPLER_YCBCR_CONVERSION_EXTENSION_NAME,
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/backends/vk/vkcontext.c:626]: We require VK_API_VERSION_1_1.

File: libnodegl/backends/vk/texture_vk.c [hunk 5]
```diff
@@ -449,6 +449,7 @@ VkResult ngli_texture_vk_wrap(struct texture *s, const struct texture_vk_wrap_pa
         return res;
 
     s_priv->image = wrap_params->image;
+    s_priv->wrapped_image = 1;
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/backends/vk/texture_vk.c:452]: Fixed.

File: libnodegl/backends/vk/pipeline_vk.c [hunk 6]
```diff
@@ -681,6 +743,21 @@ int ngli_pipeline_vk_update_uniform(struct pipeline *s, int index, const void *v
     return NGL_ERROR_GRAPHICS_UNSUPPORTED;
 }
 
+static int need_desc_set_layout_update(struct pipeline *s,
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/backends/vk/pipeline_vk.c:746]: Removed.

File: libnodegl/android_ctx.c [hunk 7]
```diff
@@ -116,6 +122,27 @@ int ngli_android_ctx_init(struct gpu_ctx *gpu_ctx, struct android_ctx *s)
     }
 #endif
 
+#if defined(BACKEND_VK)
+    if (config->backend == NGL_BACKEND_VULKAN) {
+        const struct gpu_ctx_vk *gpu_ctx_vk = (struct gpu_ctx_vk *)gpu_ctx;
+        const struct vkcontext *vk = gpu_ctx_vk->vkcontext;
+        s->has_native_imagereader_api = 1;
+        const char *required_extensions[] = {
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/android_ctx.c:130]: Done.

File: libnodegl/android_ctx.c [hunk 8]
```diff
@@ -116,6 +122,27 @@ int ngli_android_ctx_init(struct gpu_ctx *gpu_ctx, struct android_ctx *s)
...
+            VK_ANDROID_EXTERNAL_MEMORY_ANDROID_HARDWARE_BUFFER_EXTENSION_NAME,
+        };
+        for (int i = 0; i < NGLI_ARRAY_NB(required_extensions); i++) {
+            if (!ngli_vkcontext_has_extension(vk, required_extensions[i], 1)) {
+                LOG(WARNING, "required extensions %s is missing, disabling Mediacodec surface usage",
+                    required_extensions[i]);
+                s->has_native_imagereader_api = 0;
+                break;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/android_ctx.c]: Done.

File: libnodegl/backends/vk/hwmap_mediacodec_vk.c [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+        .image = mc->image,
+    };
+
+    const VkMemoryRequirements mem_reqs = {
+      .size = ahb_props.allocationSize,
+      .memoryTypeBits = ahb_props.memoryTypeBits,
+    };
+
+    const VkMemoryPropertyFlags mem_props = VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT;
+    int32_t mem_type_index = ngli_vkcontext_find_memory_type(vk, mem_reqs.memoryTypeBits, mem_props);
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/backends/vk/hwmap_mediacodec_vk.c]: Done.

File: libnodegl/backends/vk/ycbcr_sampler_vk.c [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+        return NGL_ERROR_GRAPHICS_GENERIC;
+    }
+
+    return VK_SUCCESS;
+}
+
+static int ycbcr_sampler_vk_params_eq(const struct ycbcr_sampler_vk_params *p0,
+                                      const struct ycbcr_sampler_vk_params *p1)
+{
+    return (p0->android_external_format == p1->android_external_format &&
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/backends/vk/ycbcr_sampler_vk.c:109]: Removed.

File: libnodegl/backends/vk/hwmap_mediacodec_vk.c [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+#include <sxplayer.h>
+#include <libavcodec/mediacodec.h>
+#include <vulkan/vulkan.h>
+#include <vulkan/vulkan_android.h>
+#include <android/hardware_buffer.h>
+#include <android/hardware_buffer_jni.h>
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/backends/vk/hwmap_mediacodec_vk.c:31]: Removed.

File: libnodegl/backends/vk/hwmap_mediacodec_vk.c [hunk 12]
```diff
@@ -0,0 +1,418 @@
...
+        }
+    }
+    NGLI_ALIGNED_MAT(crop_matrix) = {
+        sx,   0.0f, 0.0f, 0.0f,
+        0.0f, sy,   0.0f, 0.0f,
+        0.0f, 0.0f, 1.0f, 0.0f,
+        tx,   ty,   0.0f, 1.0f,
+    };
+    float *matrix = hwmap->mapped_image.coordinates_matrix;
+    memcpy(matrix, crop_matrix, sizeof(crop_matrix));
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/backends/vk/hwmap_mediacodec_vk.c:199]: It is planned in a later PR.

File: libnodegl/backends/vk/hwmap_mediacodec_vk.c [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+        LOG(WARNING, "ycbcr conversions do not support mipmapping: "
+            "disabling direct rendering");
+        return 0;
+    } else if (params->texture_wrap_s != NGLI_WRAP_CLAMP_TO_EDGE ||
+                params->texture_wrap_t != NGLI_WRAP_CLAMP_TO_EDGE) {
+        LOG(WARNING, "ycbcr conversions only support clamp to edge wrapping: "
+            "disabling direct rendering");
+        return 0;
+    } else if (params->texture_min_filter != params->texture_mag_filter) {
+        LOG(WARNING, "ycbcr conversions must have same min/mag filters: "
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/backends/vk/hwmap_mediacodec_vk.c]: Fixed.

File: libnodegl/backends/vk/hwmap_mediacodec_vk.c [hunk 14]
```diff
@@ -0,0 +1,418 @@
...
+        .buffer = hardware_buffer,
+    };
+
+    const VkImageCreateInfo img_info = {
+        .sType         = VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO,
+        .pNext         = &external_memory_image_info,
+        .imageType     = VK_IMAGE_TYPE_2D,
+        .extent.width  = desc.width,
+        .extent.height = desc.height,
+        .extent.depth  = 1,
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/backends/vk/hwmap_mediacodec_vk.c:250]: Done.

File: libnodegl/meson.build (hunk 3 above)
Comment (mbouron-gpsw) [reply libnodegl/meson.build:402]: Done.

File: libnodegl/backends/vk/ycbcr_sampler_vk.c (hunk 2 above)
Comment (mbouron-gpsw) [reply libnodegl/backends/vk/ycbcr_sampler_vk.c]: Fixed.

File: libnodegl/backends/vk/hwmap_mediacodec_vk.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/backends/vk/hwmap_mediacodec_vk.c]: Well, it's done in this PR.

File: libnodegl/meson.build (hunk 3 above)
Comment (cboesch-gpsw) [reply libnodegl/meson.build:402]: You've added it to the wrong release

Comment (mbouron-gpsw) [reply libnodegl/meson.build:402]: Sorry, fixed.

**Review by cboesch-gpsw — APPROVED** [review]

**Pull request merged**